Use inotify to wait for PostgreSQL startup.

hlinnaka opened the pull request:
Compute node startup time is very important. After launching PostgreSQL, use inotify to be notified immediately when it has updated the PID file, instead of polling. The polling loop had 100 ms interval so this shaves up to 100 ms from the startup time.

The loop also checked if the TCP port is open for connections, by trying to connect to it. That seems unnecessary. By the time the postmaster.pid file says that it's ready, the port should be open. Remove that check.

hlinnaka commented:
This probably doesn't work on macOS, since inotify is a Linux-only thing. I don't know if it fails at compilation time or runtime, however. We can just disable it on macOS, though, and continue relying on polling there.

Another approach would've been to just poll more aggressively, maybe every 10 ms. But if this works, this is nicer.

ololobus commented:
Thanks for looking on it! Yes, it fails to compile on macOS in CI

I googled a bit and there is a cross-platform `notify` crate that uses `inotify` on Linux, other mechanisms in BSD / macOS and falls-back to polling: https://github.com/notify-rs/notify, maybe use it?

hlinnaka commented:
> I googled a bit and there is a cross-platform `notify` crate that uses `inotify` on Linux, other mechanisms in BSD / macOS and falls-back to polling: https://github.com/notify-rs/notify, maybe use it?

A-ha, yes, that looks nice. Rewrote using 'notify'. I also split the removal for the check with connecting to the port into a separate commit, for clarity.

ololobus commented on compute_tools/src/pg_helpers.rs:
```diff
@@ -230,52 +231,83 @@ pub fn get_existing_dbs(client: &mut Client) -> Result<Vec<Database>> {
         if let Ok(file) = File::open(&pid_path) {
+            if !postmaster_pid_seen {
+                log::debug!("postmaster.pid appeared");
+                watcher.unwatch(pgdata).expect("Failed to remove dir watch");
```
I hope it will never fail, but maybe just to be more explicit about the error:
```suggestion
                watcher.unwatch(pgdata).expect("Failed to remove pgdata dir watch");
```
?

ololobus commented on compute_tools/src/pg_helpers.rs:
```diff
@@ -230,52 +231,83 @@ pub fn get_existing_dbs(client: &mut Client) -> Result<Vec<Database>> {
+                watcher.unwatch(pgdata).expect("Failed to remove dir watch");
+                watcher
+                    .watch(&pid_path, RecursiveMode::NonRecursive)
+                    .expect("Failed to add file watch");
```
And
```suggestion
                    .expect("Failed to add postmaster.pid file watch");
```
?

hlinnaka replied:
done

ololobus reviewed (approved):
Thanks!
